lib/Sema/ConstraintSystem.cpp

```diff
@@ -4574,6 +4577,57 @@ SolutionResult ConstraintSystem::salvage() {
       }
     }
 
+    if (fixes.empty())
+      return SolutionResult::forUndiagnosedError();
+
+    {
+      auto &Diags = getASTContext().Diags;
+      DiagnosticTransaction fallbackErrors(Diags);
+
+      auto diagnoseSingleError = llvm::all_of(fixes,
+          [](auto &s) {
+            return s.Fixes.size() == 1;
+          });
+
+      if (diagnoseSingleError) {
+        Diags.diagnose(
+            fixes[0].Fixes[0]->getAnchor().getStartLoc(),
+            diag::multiple_potential_errors);
+      }
+
+      bool diagnosed = false;
+      for (auto &solution : fixes) {
+        if (solution.Fixes.empty())
+          continue;
+
+        auto loc = solution.Fixes[0]->getAnchor().getStartLoc();
+
+        if (!diagnoseSingleError) {
+          Diags.diagnose(
+              solution.Fixes[0]->getAnchor().getStartLoc(),
+              diag::potential_error);
+        }
+
+        {
+          DiagnosticTransaction notes(Diags);
+
+          // FIXME: Notes do not nest, and many constraint fixes will have
+          // attached notes, but there's no way to represent a hierarchy
+          // of diagnostics.
+          notes.limitAllBehavior(DiagnosticBehavior::Note);
+
+          diagnosed |= applySolutionFixes(solution);
```

Comment:
There is a way to indicate to a fix whether it should be diagnosed as a note or not - `diagnose` accepts optional `asNote` parameter and passes it along to the FailureDiagnostic that calls `diagnoseAsNote`, default implementation currently simplify returns `false` but quite a few diagnostics are implemented as notes, so could you please consider changing default behavior `diagnoseAsNote` and augment applySolutionFixes to propagate "asNote" down to `coalesceAndDiagnose`?

Reply:
I did consider that, and went down that path at first, but `diagnoseAsNote` doesn't do the same thing as I'm doing here. `diagnoseAsNote` is implemented by various failure diagnostic types to produce candidate notes when the error is an ambiguous call or similar. For example, `MissingArgumentsFailure::diagnoseAsNote` diagnoses `diag::candidate_partial_match` which says `"candidate has partially matching parameter list %0"` at the callee declaration, because it assumes the error message has already diagnosed the call itself. That's not what I want to do here - I want to diagnose the error message but downgrade it to a note solely for the purpose of grouping it under the error message emitted above.

If we had a better way to group diagnostics into a hierarchy, I would not want these to be emitted as notes, but as nested errors.

Reply:
Hm, I wonder if we should rework some of the notes like the one you mentioned to be more descriptive and attach to the expression? The current way of handing it is not create because we loose the note when it’s attached to a declaration without a source location…